libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsMetadataUpdater.kt

```diff
@@ -0,0 +1,134 @@
+package com.mapbox.navigation.core.reroute
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.routeoptions.isEVRoute
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+// It's a temporary platform side workaround, the final solution will be on NN side.
+// See https://mapbox.atlassian.net/browse/NN-854
+internal suspend fun restoreChargingStationsMetadata(
+    originalRoute: NavigationRoute,
+    newRoutes: List<NavigationRoute>,
+    workerDispatcher: CoroutineDispatcher = Dispatchers.Default
+): List<NavigationRoute> = withContext(workerDispatcher) {
+    val newPrimaryRoute = newRoutes.first()
+    val rerouteRouteOptions = newPrimaryRoute.routeOptions
+    if (!rerouteRouteOptions.isEVRoute()) {
+        return@withContext newRoutes
+    }
+    val updatedResponseWithoutCorrectTypes = if (newPrimaryRoute.origin == RouterOrigin.Onboard) {
+        updateChargingStationsMetadataBasedOnRequestUrl(rerouteRouteOptions, newPrimaryRoute)
+    } else {
+        newPrimaryRoute.directionsResponse
+    }
+    val updatedResponse = updateChargingStationsTypes(
+        updatedResponseWithoutCorrectTypes,
+        originalRoute.getServerProvidedChargingStationsIds()
+    )
+    // TODO: optimise java memory usage creating a copy of a response NAVAND-1437
+    NavigationRoute.create(updatedResponse, rerouteRouteOptions, newPrimaryRoute.origin)
+}
+
+fun updateChargingStationsTypes(
+    response: DirectionsResponse,
+    serverProvidedStationsIds: Set<String>
+): DirectionsResponse {
+    val updatedRoutes = response.routes().map { route ->
+        val updatedWaypoints =
+            (route.waypoints() ?: emptyList()).mapIndexed { waypointIndex, waypoint ->
+                val waypointUnrecognizedProperties = waypoint.unrecognizedJsonProperties
+                val waypointMetadata = waypointUnrecognizedProperties?.get("metadata")
+                    ?.asJsonObject
+                val stationId = waypointMetadata
+                    ?.get("station_id")
+                    ?.asString
+                if (stationId != null && serverProvidedStationsIds.contains(stationId)) {
```

Comment:
Is it possible that user-provided and server-provided stations will have the same id?

Reply:
I was thinking about similar case. What if you have a route to destination and then back to the origin, and it so happen that you need to charge on the same charging station 2 times. One time it's user provided and the second time server added the same charging station. In this case all other fields will be the same

Reply:
TODO: create a ticket for that case

Reply:
https://mapbox.atlassian.net/browse/NAVAND-1441